topology/probes/ovs_of.go

```diff
@@ -0,0 +1,503 @@
+
+// Event is an event as monitored by ovs-ofctl monitor <br> watch:
+type Event struct {
+	Rule   *Rule  // The rule modified
+	Date   int64  // the date of the event
+	Action string // the action taken
+	Bridge string // The bridge whtere it ocured
+}
+
+// ProtectCommas substitute commas with semicolon
+// when inside parenthesis
+func ProtectCommas(line string) string {
+	work := []rune(line)
+	var braces = 0
+	for i, word := range work {
+		switch word {
+		case '(':
+			braces++
+		case ')':
+			braces--
+		case ',':
+			if braces > 0 {
+				work[i] = ';'
+			}
+		}
+	}
+	return string(work)
+}
+
+// fillIn is a utility function that takes a splitted rule line
+// and fills a Rule/Event structure with it
+func fillIn(components []string, rule *Rule, event *Event) {
+	for _, component := range components {
+		keyvalue := strings.SplitN(component, "=", 2)
+		if len(keyvalue) == 2 {
+			key := keyvalue[0]
+			value := keyvalue[1]
+			switch key {
+			case "event":
+				if event != nil {
+					event.Action = value
+				}
+			case "actions":
+				rule.Actions = value
+			case "table":
+				table, err := strconv.ParseInt(value, 10, 32)
+				if err == nil {
+					rule.Table = int(table)
+				} else {
+					logging.GetLogger().Errorf("Error while parsing table of rule: %s", err.Error())
+				}
+			case "cookie":
+				v, err := strconv.ParseUint(value, 0, 64)
+				if err == nil {
+					rule.Cookie = v
+				} else {
+					logging.GetLogger().Errorf("Error while parsing cookie of rule: %s", err.Error())
+				}
+			}
+		}
+	}
+
+}
+
+// ParseEvent transforms a single line of ofctl monitor :watch
+// in an event. The string must be terminated by a "\n"
+// Protected commas will be replaced.
+func ParseEvent(line string, bridge string, prefix string) (Event, error) {
+	var result Event
+	var rule Rule
+
+	if line[0] != ' ' {
+		return result, errors.New("No event")
+	}
+	if strings.ContainsRune(line, '(') {
+		line = ProtectCommas(line)
+	}
+	components := strings.Split(line[1:len(line)-1], " ")
+	fillIn(components, &rule, &result)
+	if len(components) < 2 {
+		return result, errors.New("Rule syntax")
+	}
+	if len(components) > 4 {
+		rule.Filter = components[len(components)-2]
+	} else if len(components) == 4 && rule.Actions == "" {
+		rule.Filter = components[len(components)-1]
+	}
+	result.Rule = &rule
+	fillUUID(&rule, prefix)
+	result.Date = time.Now().Unix()
+	result.Bridge = bridge
+	return result, nil
+}
+
+// Generates a unique UUID for the rule
+// prefix is a unique string per bridge using bridge and host names.
+func fillUUID(rule *Rule, prefix string) {
+	id := prefix + rule.Filter + "-" + string(rule.Table) + "-" + string(rule.Cookie)
+	u, err := uuid.NewV5(uuid.NamespaceOID, []byte(id))
+	if err == nil {
+		rule.UUID = u.String()
+	}
+}
+
+// ParseRule transforms a single line of ofctl dump-flow in a rule.
+// The line DOES NOT include the terminating newline. Protected commas will be replaced.
+func ParseRule(line string) (Rule, error) {
```

Comment:
Yes because OVS format are different for each kind of tool it is hard to share it between modules.